File: TicTacToe.py
'''
Simple Tic-Tac-Toe Game in Python
Written by Matthew D'Arcy
'''

from random import randint
from IPython.display import clear_output

p1_name = ''
p2_name = ''
p_turn = {'name': '', 'symbol': ''}
game_over = False
move_count = 0
spaces = {
    'm1': 1,
    'm2': 2,
    'm3': 3,
    'm4': 4,
    'm5': 5,
    'm6': 6,
    'm7': 7,
    'm8': 8,
    'm9': 9,
}


# Randomize player symbol and turn order

def rand(p1,p2):
    
    p1_first = randint(0,1)
    
    if p1_first == 0:
        
        return {'name': p2,'symbol':'O'}
    
    elif p1_first == 1:
        
        return {'name': p1,'symbol':'X'}


# Change turn between players

def swap_player():
    
    global p_turn
    
    global p1_name
    
    global p2_name
    
    if p_turn['symbol'] == 'X':
        
        p_turn = {'name': p2_name, 'symbol': 'O'}
        
    elif p_turn['symbol'] == 'O':
        
        p_turn = {'name': p1_name, 'symbol': 'X'}
    
    if game_over == False:
        clear_output()


# Welcome routine collecting user information

def welcome():
    
    print 'Welcome to tic-tac-tsssss!'
    
    print "Player 1, X's, please enter your name:"
    
    global p1_name
    
    p1_name = raw_input()
    
    print 'Welcome to tic-tac-tsssss, {x}'.format(x = p1_name)
    
    print 
    
    global p2_name
    
    print "Player 2, O's, please enter your name:"
    
    p2_name = raw_input()
    
    print 'Welcome to tic-tac-tsssss, {x}'.format(x = p2_name)
    
    print
    
    print 'Randomizing first player...'
    
    print
    
    global p_turn
    
    p_turn = rand(p1_name, p2_name)


# Print current game board

def board():
    
    print(" {s1} | {s2} | {s3}".format(s1 = spaces['m1'], s2 = spaces['m2'], s3 = spaces['m3']))
    
    print("---+---+---")
    
    print(" {s4} | {s5} | {s6}".format(s4 = spaces['m4'], s5 = spaces['m5'], s6 = spaces['m6']))
    
    print("---+---+---")
    
    print(" {s7} | {s8} | {s9}".format(s7 = spaces['m7'], s8 = spaces['m8'], s9 = spaces['m9']))


# To begin game again, reinitialize appropriate parameters

def start_game():
    
    global game_over
    
    global move_count
    
    global spaces
    
    global p_turn
    
    game_over = False
    
    move_count = 0
    
    spaces = {
        'm1': 1,
        'm2': 2,
        'm3': 3,
        'm4': 4,
        'm5': 5,
        'm6': 6,
        'm7': 7,
        'm8': 8,
        'm9': 9,
    }
    
    p_turn = {'name': '', 'symbol': ''}
    
    welcome()
    
    while game_over == False:
    
        board()
    
        get_input()
    
        game_over_check()
    
        swap_player()


# Check for gameover conditions
    
def game_over_check():
    
    global spaces
    
    global game_over
    
    global p_turn
    
    global move_count
   
    move_count += 1
    
    if ((spaces['m1'] == spaces['m2'] == spaces['m3']) or 
        (spaces['m4'] == spaces['m5'] == spaces['m6']) or 
        (spaces['m7'] == spaces['m8'] == spaces['m9']) or 
        (spaces['m1'] == spaces['m4'] == spaces['m7']) or 
        (spaces['m2'] == spaces['m5'] == spaces['m8']) or 
        (spaces['m3'] == spaces['m6'] == spaces['m9']) or 
        (spaces['m1'] == spaces['m5'] == spaces['m9']) or 
        (spaces['m3'] == spaces['m5'] == spaces['m7'])):
        
            clear_output()
        
            board()
            
            print '{x} wins!'.format(x = p_turn['name'])
            
            game_over = True
            
    if (move_count == 9):
        
        print 'Game ended in a tie!'
        
        game_over = True
    
    if game_over == True:
        decision_made = False
        
        while decision_made == False:

            user_input = raw_input("Would you like to play again? (Please enter 'y' or 'n'): ")

            if user_input == 'y':
                
                clear_output()
                
                start_game()
                
                decision_made = True
                
            elif user_input == 'n':
                
                raise SystemExit('Thanks for playing!')
                

# Verify is an integer, apply to user input for space selection

def is_int(x):
    
    try: 
        
        int(x)
        
        return True
    
    except ValueError:
        
        return False
    

# Get input from user and run checks for valid selection

def get_input():
    
    user_input_int = ''
    
    good_input = False
    
    global spaces
    
    while good_input == False:
        
        user_input_str = raw_input('{p_name}, please pick a free space from 1 to 9 to place {p_symbol}: '.format(p_name = p_turn['name'], p_symbol = p_turn['symbol']))
        
        if is_int(user_input_str) == True:
            
            user_input_int = int(user_input_str)
            
        m_name = 'm' + user_input_str
    
        if ((is_int(user_input_int) == False) or (user_input_int < 1) or (user_input_int > 9)):
            
            user_input = raw_input('{p_name}, please pick a free space from 1 to 9 to place {p_symbol}: '.format(p_name = p_turn['name'], p_symbol = p_turn['symbol']))
            
        elif ((is_int(user_input_int) == True) and 
              
              (user_input_int >= 1) and 
              
              (user_input_int <= 9) and 
              
              (is_int(spaces[m_name]))):
            
            good_input = True
            
            spaces[m_name] = p_turn['symbol']


# main loop

start_game()


    

File: TicTacToe.c
/*
Simple Tic-Tac-Toe Game in C
Written by Matthew D'Arcy
*/

#include<stdio.h>
#include<string.h>
#include<stdlib.h>
#include<time.h>
#include<unistd.h>
#include<stdbool.h>

#define name_limit 64
#define TRUE 1
#define FALSE 0

struct playerdata	
{
	char P[64];
	char marker;	
};	

struct playerdata player1;
struct playerdata player2;
int total_choices[9] = {0,0,0,0,0,0,0,0,0};
char chararray[3][3] = { {'1','2','3'} , {'4','5','6'} , {'7','8','9'} };
int straight = 0, goplayer = 0, p1 = 0, p2 = 0, p3 = 0;
int gameover = FALSE;
int last, row, col;
int isdraw = 0;
char replay;
char null;
int play = TRUE;

/* Display tic-tac-toe board to user */
void show_board()
{	
	printf("\n %c | %c | %c\n",chararray[0][0],chararray[0][1],chararray[0][2]);
	printf("---+---+---\n");
	printf(" %c | %c | %c\n",chararray[1][0],chararray[1][1],chararray[1][2]);
	printf("---+---+---\n");
	printf(" %c | %c | %c\n\n",chararray[2][0],chararray[2][1],chararray[2][2]);
}

/* Upon game-end condition, ask to play again, if so reinitialize, otherwise quit */
void endgame()
{
	show_board();	
	gameover = TRUE;	
	if(goplayer == 1 && isdraw == 0)
	{
		printf("Game Over!\nThe winner is %s",player2.P);
	}
	else if (goplayer == 2 && isdraw == 0)
	{
		printf("Game Over!\nThe winner is %s",player1.P);
	}
	else
	{
		printf("The game is a draw!\n\r");
	}
	sleep(2);
	printf("Play again? (y for yes, anything else to end)\n\r");
	scanf("%c",&replay);
	while(getc(stdin) != '\n');
	if(replay == 'y' || replay == 'Y')
	{		
		chararray[0][0] = '1';
		chararray[0][1] = '2';
		chararray[0][2] = '3';
		chararray[1][0] = '4';
		chararray[1][1] = '5';
		chararray[1][2] = '6';
		chararray[2][0] = '7';
		chararray[2][1] = '8';
		chararray[2][2] = '9';
		memset(total_choices,0,sizeof(total_choices));
		straight = 0, goplayer = 0, p1 = 0, p2 = 0, p3 = 0;
		isdraw = 0;	
	}
	else
	{		
		printf("\nGoodbye!\n\n\r");		
		exit(0);
	}
}

/* Check for player win cases */
void gamecheck()
{	
	total_choices[p3] = last;
	p3++;
	if ((chararray[0][0] == chararray[1][1] && chararray[1][1] == chararray[2][2]) || (chararray[0][2] == chararray[1][1] && chararray[1][1] == chararray[2][0])) 
	{
		endgame();
	}
	else
	{	
		for(straight = 0; straight <=2; straight++)
		{
			if((chararray[straight][0] == chararray[straight][1] && chararray[straight][1] == chararray[straight][2]) || (chararray[0][straight] == chararray[1][straight] && chararray[1][straight] == chararray[2][straight]))
			{
				endgame();
			}
		}
		if (p3 > 8)
		{
			printf("p3 is %d",p3);
			isdraw = 1;
			endgame();
		}	
	}	
}

/* Randomly choose players' order and markers */
void first()
{
	sleep(1);
	printf("Randomizing...\n\n\r");	
	sleep(2);
	srand(time(NULL));
	int first = (rand()%4);	
	switch(first)
	{
		case 0:
			printf("Turn: %sYour symbol is X\n\n\r",player1.P);
			player1.marker = 'X';
			player2.marker = 'O';
			goplayer = 1;
			break;
		case 1:
			printf("Turn: %sYour symbol is O\n\n\r",player1.P);
			player1.marker = 'O';
			player2.marker = 'X';			
			goplayer = 1;
			break;
		case 2:
			printf("Turn: %sYour symbol is X\n\n\r",player2.P);
			player2.marker = 'X';	
			player1.marker = 'O';		
			goplayer = 2;
			break;
		case 3:
			printf("Turn: %sYour symbol is O\n\n\r",player2.P);
			player2.marker = 'O';	
			player1.marker = 'X';		
			goplayer = 2;
			break;
	}
	sleep(1);
}

int clean_user_input()
{
	while(getchar() != '\n');
	return(1);
}

/* Get user input for space selection and check validity */
void choose_space()
{	
	do
	{
		printf("(single digit 1 through 9): ");
	} while ((((scanf("%d%c", &last, &null) != 2) || null != '\n') && clean_user_input()) || last < 1 || last > 9);
	
	int j;	
	for(j=0;j<9;j++)
	{
		if(total_choices[j]==last)
		{
			printf("Space already chosen!\nPlease choose again: ");
			choose_space();
		}
	}	
	row = (last-1)/3;
	col = (last-1)%3;	
}

/* Place X or O where user has selected */
void place()
{
	switch(goplayer)
	{
		case 1:
			printf("Enter the number of your field of choice, %s", player1.P);
			choose_space();
			chararray[row][col] = player1.marker;		
			goplayer = 2;
			p1++;
			gamecheck();			
			break;
		case 2:
			printf("Enter the number of your field of choice, %s", player2.P);	
			choose_space();
			chararray[row][col] = player2.marker;
			goplayer = 1;
			p2++;
			gamecheck();
			break;
	}		
	if(gameover == FALSE)
	{	
		show_board();
	}
}

/* Main routine */
int main()
	{
		while(play == TRUE)
		{	
			gameover = FALSE;
			printf("\n\nWelcome to Tic-Tac-Toe!\n\n\n");
			printf("Please submit name of player 1: ");							
			fgets(player1.P,name_limit,stdin);	
			printf("Welcome %s\n", player1.P);
			printf("Please submit name of player 2: ");		
			fgets(player2.P,name_limit,stdin);
			printf("Welcome %s\n", player2.P);
			first();	
			show_board();
			while(gameover == FALSE)
			{
				place();
			}			
		}
		return(0);	
	}
